### ISSUE TITLE
Hyperopt-loss: ProfitSquare with Sortino-Spring Efficiency (PSSE)

### ISSUE THREAD

**Issue by keithorange** [freqtrade/freqtrade#8810]
# 📈 Introducing ProfitSquare with Sortino-Spring Efficiency (PSSE): The Evolution in Portfolio Optimization 📊

## Environment Requirements
- Operating System: Cross-platform
- Python Version: Python 3.7+ (`python -V`)
- CCXT version: 1.60.92+ (`pip freeze | grep ccxt`)
- Freqtrade Version: 2023.06 (`freqtrade -V`)

## The Genesis of PSSE 🧪

Dear quantitative developers and trading strategists, it's time to take a leap forward in portfolio optimization. We present the **ProfitSquare with Sortino-Spring Efficiency (PSSE)**, a cutting-edge loss function forged at the intersection of financial acumen and algorithmic prowess. PSSE is designed to address the intricacies of market dynamics and empowers your strategies to optimize for rewards, judiciously considering risks and drawdowns.

## Core Pillars of PSSE 🏛

1. **Quadratic Profit**: Employs the square of profits to accentuate performance. This non-linear approach facilitates the exploration of investment spaces beyond linear boundaries.

2. **Sortino Ratio**: Modifies the Sharpe ratio to focus on downside risk, effectively functioning as the navigation system for the portfolio.

3. **Hybrid Drawdown Penalty**: A meticulously engineered combination of geometric and harmonic drawdown penalties. It serves as the stabilizing force, akin to the dampening system of a high-performance vehicle.

Together, these elements synthesize into:
```
PSSE = profit^2 + (profit * Sortino ratio) / hybrid drawdown penalty
```

## Inside the Codebase 🧩

For the discerning developer, let's dissect the primary class: `ProfitSquareSortinoSpringEfficiencyHyperOptLoss`. It integrates PSSE as the loss function for hyperparameter optimization.

```
def calculate_average_drawdown(drawdown_df):
    """
    Calculate average drawdown based on a DataFrame containing drawdown information.

    Parameters:
    - drawdown_df: DataFrame containing columns 'high_value', 'cumulative' and 'drawdown_relative'

    Returns:
    - Tuple of average absolute drawdown and average relative drawdown
    """
    # Copy input DataFrame to avoid modifying the original
    dd_df = drawdown_df.copy()

    # Calculate drawdowns
    dd_df['calculated_drawdowns'] = dd_df['high_value'] - dd_df['cumulative']

    # Calculate the average drawdown (absolute)
    average_drawdown = dd_df['calculated_drawdowns'].mean()

    # Calculate the average drawdown (relative)
    average_drawdown_relative = dd_df['drawdown_relative'].mean()

    return average_drawdown, average_drawdown_relative


class ProfitSquareSortinoSpringEfficiencyHyperOptLoss(IHyperOptLoss):
    """
    Custom loss function for hyperopt optimization combining quadratic profits, Sortino ratio,
    and hybrid drawdown penalty. The loss function is designed to optimize the trading strategies
    by prioritizing higher profits and lower drawdowns, while adjusting for risk through Sortino ratio.
    """

    @staticmethod
    def hyperopt_loss_function(results: DataFrame, trade_count: int,
                               min_date: datetime, max_date: datetime,
                               config: Config, *args, **kwargs) -> float:
        """
        Calculate the hyperopt loss based on trade results.

        Parameters:
        - results: DataFrame containing the trading results
        - trade_count: int representing the number of trades
        - min_date: datetime object representing the minimum date in the dataset
        - max_date: datetime object representing the maximum date in the dataset
        - config: Config object containing the configuration

        Returns:
        - The calculated loss as a float
        """

        try:
            # Total profit
            profit = results['profit_abs'].sum()

            # Sortino ratio
            sortino_ratio = calculate_sortino(results, min_date, max_date, config['dry_run_wallet'])

            # Underwater DataFrame (max value to date, drawdown, drawdown relative)
            drawdown_df = calculate_underwater(
                results,
                value_col='profit_abs',
                starting_balance=config['dry_run_wallet']
            )

            # Calculate Max Drawdown
            max_drawdown = abs(min(drawdown_df['drawdown']))

            # Calculate Average Drawdown (absolute and relative)
            avg_max_drawdown, _ = calculate_average_drawdown(drawdown_df)

            # PSSE Loss Function
            if profit > 0:
                # Sortino ratio requires at least 2 trades for meaningful results
                if trade_count < 2:
                    sortino_ratio = 1

                # Geometric mean of drawdown penalties
                geometric_drawdown_penalty = math.sqrt((1 + max_drawdown) * (1 + avg_max_drawdown))

                # Harmonic mean of drawdown penalties
                harmonic_drawdown_penalty = 2 / \
                    ((1 / (1 + max_drawdown)) + (1 / (1 + avg_max_drawdown)))

                # Hybrid drawdown penalty (geometric mean combined with harmonic mean)
                hybrid_drawdown_penalty = math.sqrt(
                    geometric_drawdown_penalty * harmonic_drawdown_penalty)

                # Combine quadratic profit with Sortino ratio and hybrid drawdown penalty
                loss = -1 * ((profit ** 2) + (profit * sortino_ratio) / hybrid_drawdown_penalty)
            else:  # profit < 0
                loss = -profit

            return loss

        # Handle any exception that may occur during loss calculation
        except:
            traceback.print_exc()
            return 1000000

```

### Notable Aspects:

- Utilizes both the `Sortino ratio` and a `hybrid drawdown penalty`, optimizing for high returns while safeguarding against significant drawdowns.
- The hybrid drawdown penalty amalgamates geometric and harmonic means, encapsulating both the magnitude and the persistence of drawdowns.
- Cognizant of the potential for overfitting, PSSE ensures that the model parameters are not over-optimized to historical data.

## Practical Application 🛠

Use PSSE as a custom loss function for hyperparameter optimization in your trading algorithms. It can prove invaluable in periods of market volatility, ensuring that your strategies remain robust, maximizing profits while controlling risks and drawdowns.

## PSSE - Your Intelligent Co-pilot 🤖

In an era where financial markets are more tumultuous than ever, PSSE stands as an intellectual accomplice in your pursuit of profit. Its mathematical rigor coupled with empirical substantiation makes it an indispensable tool in the quant developer's arsenal.

Contributions, feedback, and discussions are highly valued. Together, let's refine the frontiers of quantitative finance.

## Keywords
ProfitSquare, Sortino-Spring Efficiency, Quantitative Finance, Portfolio Optimization, Risk Management
```

**Comment by xmatthias** [comment]
you should consider converting this to a pull request.

I understand this can be challenging if you never did, but it's not that difficult, and pretty well documented in various parts, for example:
[create a pull request](https://opensource.com/article/19/7/create-pull-request-github)


A guide for the github PR workflow can be found [here](https://guides.github.com/introduction/flow/).

---

We'll not be considering contributions in issues - as that's only causing work on our side, and prevents the usage of proper review methods, like "line comments" to highlight exactly where there are questions, doubts or necessary changes.

Thanks for your understanding.

**Comment by xmatthias** [comment]
Closing this, as the PR was abandoned (so the assumption is  there's no further interest).
